Добалены JUnit тесты

ml-rdv opened the pull request:
+ изменён метод extendArray()

ml-rdv committed:
Добалены JUnit тесты
+ изменён метод extendArray()

ml-rdv committed:
Добалены JUnit тесты для класса MyQueue

ml-rdv committed:
Добалены JUnit тесты для класса MyStack

Adgam commented on src/DataStructures/Stack/MyStack.java:
```diff
@@ -21,11 +21,17 @@ public void push(int element) {
     }
 
-    public int pop() {
+    public int pop() throws IndexOutOfBoundsException {
```
IndexOutOfBoundsException является unchecked исключением, это значит, что в секцию throws можно не добавлять это исключение

Adgam commented on src/DataStructures/Stack/MyStackTesting.java:
```diff
@@ -1,19 +1,45 @@
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
 public class MyStackTesting {
```
Обычно классы тестов называют по шаблону SomethingTest, т.к. testing это всё же "тестирование", а test - конкретный тест

Adgam commented on src/DataStructures/Stack/MyStackTesting.java:
```diff
@@ -1,19 +1,45 @@
+    }
+
+    @Test
+    void should_peek_element_from_stack() {
```
не хватает ещё одного assert, в чём отличие pop от peek?

Adgam commented on src/DataStructures/Queue/MyQueueTesting.java:
```diff
@@ -1,27 +1,60 @@
+    }
+
+    @Test
+    void should_peek_element_from_queue() {
```
аналогично со стеком, какой проверки не хватает?

Adgam commented:
Не хватает тестов на push и size в MyStack.

Дополнительно прочитай плиз про .gitignore, посмотри зачем он нужен и что в него писать. 
А то файлы по типу hw1_2.iml не должен попадать в коммит

ml-rdv committed:
Добалены JUnit тесты для класса MyBidirectionalLinkedList + изменён метод delete (добавлена переменная tail, которая ссылается на конец листа)

ml-rdv commented:
@Adgam По поводу push в MyStack поняла, но как же проверка на size? 
Если не ошибаюсь, ты говорил, что в стеке мы не можем узнать размер с помощью функции size, там нет такого понятия

Adgam commented on src/DataStructures/BidirectionalLinkedList/MyBidirectionalLinkedList.java:
```diff
@@ -14,30 +15,34 @@ public void insertAtHead(int number) throws CloneNotSupportedException {
+            Node currentValue = head;
+            Node newElement = new Node();
+            newElement.setData(number);
+            for (int i = 0; i < size; i++) {
```
всё ещё линейная сложность вставки в конец списка, если тебе известен последний элемент, то зачем проходить по всем элемента с первого, когда ты берёшь сразу последний и переставляешь связи

Adgam commented:
В MyBidirectionalLinkedList не вижу тестов на добавление элементов

ml-rdv committed:
Добалены JUnit тесты для класса MyBinarySearchTree

ml-rdv committed:
Добален метод size() + JUnit тесты для push() и size()

ml-rdv committed:
1. убрала секцию throws IndexOutOfBoundsException
2. Переименовала классы по шаблону SomethingTest
3. Добавила тесты на pop и peek в стеке
4. Добавила тесты на dequeue и peek в очереди
5. Переписала метов insertAtEnd в MyBidirectionalLinkedList (линейная сложность)
6. В MyBidirectionalLinkedList  добавила тесты на добавление элементов

ml-rdv commented:
@Adgam  fixed ✅

Adgam commented on src/DataStructures/BidirectionalLinkedList/Node.java:
```diff
@@ -5,6 +5,23 @@ public class Node {
     private Node next;
 
+    public Node() {
+        previous = null;
```
null в конструкторах можно не присваивать, т.к. поля ссылочного типа по умолчанию == null

Adgam commented on src/DataStructures/Stack/MyStackTest.java:
```diff
@@ -71,4 +71,32 @@ void should_throw_IndexOutOfBoundsException_when_peek() {
     }
+
+    @Test
+    void check_size_after_pop_in_stack() {
```
объедини с  should_pop_element_from_stack, просто перенеси проверку на размер отсюда в should_pop_element_from_stack

Adgam commented on src/DataStructures/Stack/MyStackTest.java:
```diff
@@ -71,4 +71,32 @@ void should_throw_IndexOutOfBoundsException_when_peek() {
+    }
+
+    @Test
+    void check_size_after_peek_in_stack() {
```
объедини с  should_peek_element_from_stack, просто перенеси проверку на размер отсюда в should_peek_element_from_stack

ml-rdv committed:
1. Убрала присвоение значения null в конструкторе класса Node
2. Объединила тесты на проверку size после методов pop и peek в классе MyStackTest и после методов dequeue и peek в классе MyQueueTest

ml-rdv commented:
@Adgam fixed ✅

Adgam commented on src/DataStructures/Queue/MyQueueTest.java:
```diff
@@ -25,8 +25,10 @@ void should_dequeue_element_from_queue() {
-
         Assertions.assertEquals(expectedElement, 2);
+
+        var expectedElement2 = myQueue.size();
```
expectedSize

Adgam commented on src/DataStructures/Queue/MyQueueTest.java:
```diff
@@ -45,8 +47,10 @@ void should_peek_element_from_queue() {
-
         Assertions.assertEquals(expectedElement, 2);
+
+        var expectedElement2 = myQueue.size();
```
expectedSize

ml-rdv committed:
Renamed expectedElement2 to expectedSize

ml-rdv commented:
@Adgam fixed ✅

ml-rdv committed:
1. Добавлены тесты на вставку элемента в бинарное дерево поиска
2. Изменён метот insert
3. Добавлен метод insertNode
4. Изменён метот removeNodeWithoutChild, removeNodeWithoutChild, removeNodeWithRightChild
5. Добавлен метод removeNode

ml-rdv committed:
Исправлена ошибка
При удалении элемента, у которого нужно найти в правом поддереве минимальный элемент и переместить его на место удаляемого узла, а потомков перемещенного элемента прикрепить к элементу выше

ml-rdv committed:
Корректировки

ml-rdv committed:
Корректировки

ml-rdv committed:
Добавлен тест

ml-rdv committed:
Добавлена реализация хэш таблицы

ml-rdv committed:
Корректировки

ml-rdv committed:
Изменен класс MyHashTable
Сложность вставки, удаления и поиска элемента = О(1)

ml-rdv committed:
Корректировки

ml-rdv committed:
Корректировки

Adgam commented on src/DataStructures/BinarySearchTree/MyBinarySearchTree.java:
```diff
@@ -1,80 +1,78 @@
         if (node.getData() > number) {
             if (node.getLeft() == null) {
-                return node;
+                insertNode(node, number, true);
```
Я предлагаю избавиться от метода insertNode по двум причинам.
1. Использование флагов в сигнатуре метода в большинстве случаев является запахом кода и приводит к усложнению понимания кода (прочитай плиз это https://martinfowler.com/bliki/FlagArgument.html) 
2. Инкремент size по логике всё же должен быть в addNode, да и сама часть с добавлением ребёнка в ноду не очень претендует на перемещение в отдельный метод. 

Можешь привести аргументы, почему insertNode всё же должен существовать?

Такой вариант совершенно нормальный и визуально сразу идёт понимание добавления нового элемента.
```java
    if (node.getLeft() == null) {
        node.setLeft(new Node(number));
        size++;
        return;
    }
```

Если же всё таки прям хочется вынести метод, то я бы это сделал двумя вариантами:
1. Разделение на 2 метода, как предлагается в статье
```java
  private void insertLeftNode(Node node, int number) {
      Node newNode = new Node(number, null, null);
      node.setLeft(newNode);
      size++;
  }
```
2.  Через функциональное программирование, как по мне в данном кейсе не уместно
```java
insertNode(number, node::setLeft);
insertNode(number, node::setRight);
.............

private void insertNode(int number, Consumer<Node> setChild) {
    Node newNode = new Node(number, null, null);
    setChild.accept(newNode);
    size++;
}
```

Adgam commented on src/DataStructures/BinarySearchTree/MyBinarySearchTree.java:
```diff
@@ -1,80 +1,78 @@
         if (node == null) {
-            return stringBuilder;
+            return;
         } else {
```
else можно убрать и просто оставить 
```java
stringBuilder.append(node.getData()).append(' ');
```

Adgam commented on src/DataStructures/BinarySearchTree/MyBinarySearchTree.java:
```diff
@@ -94,13 +92,14 @@ private Node recursiveSearch(Node node, int number) {
         Node parent = findParent(root, number);
         Node childToRemove;
-        boolean itIsLeftChild = false;
+        boolean itIsLeftChild;
```
Попробуй избавиться от флага itIsLeftChild. После рефакторинга методов removeNodeWith... в них осталось по одной строчки, так ли они вообще нужны, почему сразу не вызвать removeNode без этих посредников?

Если ты используешь методы removeNodeWith... для добавления понимания какой вид удаления используется в том или ином случае, то тут могу предложить следующий подход.
```java
// Данную конструкцию
if (childToRemove.getLeft() == null && childToRemove.getRight() == null) {
    //removeNode()
} else if (childToRemove.getLeft() != null && childToRemove.getRight() == null) {
     //removeNode()
}

//можно заменить на
boolean hasNotChildren = childToRemove.getLeft() == null && childToRemove.getRight() == null;
boolean hasLeftChild = childToRemove.getLeft() != null && childToRemove.getRight() == nul;

if (hasNotChildren) {
     //removeNode()
} else if (hasLeftChild) {
     //removeNode()
}

// в принципе можешь пойти дальше и вынести определение признаков детей в методы ноды
if (childToRemove.hasNotChildren()) {
     //removeNode()
} else if (childToRemove.hasLeftChild()) {
     //removeNode()
}
```

Adgam commented on src/DataStructures/BinarySearchTree/MyBinarySearchTreeTest.java:
```diff
@@ -5,6 +5,36 @@
+    }
+
+    @Test
+    void should_not_insert_element_to_binarySearchTree() {
```
скорей should_not_insert_same_elements_to_binarySearchTree

ml-rdv committed:
Изменен метод addNode
Убрала метод insertNode
Изменен метод remove

ml-rdv committed:
В методе remove разделила removeNode на removeLeftNode и removeRightNode, чтобы избавиться от передачи флага itIsLeftChild

ml-rdv committed:
Корректное удаление корня + добавлены тесты

ml-rdv committed:
Форматирование

ml-rdv committed:
Рефакторинг класса MyHashTable, MyHashTableTest
1. Класс Pair изменен по аналогии связного списка
2. Изменён метод insert, remove, find

Adgam commented on src/DataStructures/BinarySearchTree/MyBinarySearchTree.java:
```diff
@@ -85,121 +85,88 @@ private Node recursiveSearch(Node node, int number) {
-        } else if (parent.getRight() != null && parent.getRight().getData() == number) {
-            childToRemove = parent.getRight();
-            itIsLeftChild = false;
+        if (root.getData() == number) {
```
возможен выброс исключения NullPointerException, добавь тест, где ты сразу вызываешь удаление без добавление элементов и посмотри что будет

Adgam commented on src/DataStructures/BinarySearchTree/MyBinarySearchTree.java:
```diff
@@ -85,121 +85,88 @@ private Node recursiveSearch(Node node, int number) {
+            removeNode(root, number);
         }
+        size = 0;
+        checkSize(root);
```
выглядит очень не оптимально, может есть более простой способ изменять кол-во элементов при удаление? А то тут на каждое удаление приходится рекурсивно обходить всё дерево для подсчёта.

Adgam replied:
а ещё название путает, это же не проверка размера, а скорей перерасчёт или просто расчёт размера

Adgam commented on src/DataStructures/HashTable/MyHashTable.java:
```diff
@@ -1,75 +1,73 @@
+        if (key < 0) {
+            key = key * -1;
+        }
+        return key % arr.length;
```
В данном случае позволительно использовать Math.abs()

Adgam commented on src/DataStructures/HashTable/MyHashTable.java:
```diff
@@ -1,75 +1,73 @@
+            pair = pair.getNext();
+        }
+        if (pair == null) {
+            throw new NullPointerException("Key " + key + " is absent.");
```
выбрасывать здесь NullPointerException не очень правильно, по хорошему можно создать своё исключение с именованием, которое явно покажет, что не так произошло, а пока можно оставить общее исключение RuntimeException

ml-rdv committed:
Рефакторинг
1. Добавлена обработка NullPointerException в метод remove класса MyBinarySearchTree
2. Изменен пересчет size при удалении в MyBinarySearchTree
3. Использование Math.abs()
4. Переименованы тесты
